### PR TITLE
Allow running clang-tidy in parallel

### PR DESCRIPTION
**Pull request by nilsdeppe**
## Proposed changes

- Add a `.clang-tidy` file
- Search for and use `run-clang-tidy` when it exists
- Add support for running clang-tidy on multiple threads using the `NUM_THREADS=` argument to the make command
- Run clang-tidy on 2 threads in CI

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [ ] Code
- [ ] Documentation
- [x] Build system
- [x] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

File: tools/ClangTidyAll.sh [hunk 1]
```diff
@@ -8,8 +8,53 @@ ALL_SOURCE_FILES=`find \
...
+            break
+        fi
+    fi
 done
+
+EXIT_CODE=0
+
+# Check for run-clang-tidy, then run-clang-tidy-LLVM_VERSION, and
+# if we can't find those, then just loop over the files one at a time.
+if command -v run-clang-tidy > /dev/null 2>&1; then
```
Comment (wthrowe) [inline tools/ClangTidyAll.sh]: Might be good to check `run-clang-tidy.py` explicitly as well.  You can turn the repeated checks into a loop like
```
for run-clang-tidy in run-clang-tidy run-clang-tidy.py "${RUN_CLANG_TIDY_BIN}" ;
do
    if command -v "${run-clang-tidy}" >/dev/null 2>&1; then
etc.
```

I guess you'll need a flag to stop when you've found a working one.  Same for the other file.

**Review by wthrowe — CHANGES_REQUESTED** [review]

File: .clang-tidy [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+-bugprone-exception-escape,
+# complains about Catch macros having infinite loops,
+-bugprone-infinite-loop,
+-bugprone-macro-parentheses,
+# many static variables we use do not throw and if they do we,
+# want to terminate anyway,
+-cert-err58-cpp,
+-cert-msc51-cpp,
+-cert-msc32-c,
+# checks for incorrectly implemented self-assignment
```
Comment (markscheel) [inline .clang-tidy]: Need comma at end of line?  (or change line 5 above that says every comment line must end in a comma)

File: tools/ClangTidyAll.sh [hunk 3]
```diff
@@ -8,8 +8,53 @@ ALL_SOURCE_FILES=`find \
...
+fi
+
+# Try to find run-clang-tidy-LLVM_VERSION in the path.
+RUN_CLANG_TIDY_BIN=
+for dir in $(echo ${PATH} | tr ':' '\n');
 do
-    printf "\n\nChecking file: $file\n"
-    make clang-tidy FILE=$file
+    if [ -d $dir ]; then
+        RUN_CLANG_TIDY_BIN=`find $dir -name run-clang-tidy* | head -n 1`
```
Comment (markscheel) [inline tools/ClangTidyAll.sh:34]: Good! The star is important.  Sometimes there is run-clang-tidy-LLVM_VERSION.py and not run-clang-tidy.py (but maybe this is a problem only for old LLVM).

File: tools/ClangTidyAll.sh (hunk 1 above)
Comment (markscheel) [reply tools/ClangTidyAll.sh]: Agree on checking run-clang-tidy.py and putting into a loop with a `break` and a flag if you find one (the flag so you know whether to run 'make clang-tidy').

**Review by markscheel — CHANGES_REQUESTED** [review]
Minor changes.

I believe this PR will make it easier for ninja users to run clang-tidy.

**Comment by nilsdeppe** [comment]
I've finally pushed fixups... Any more feedback would be greatly appreciated!

File: tools/ClangTidyAll.sh [hunk 4]
```diff
@@ -42,14 +42,17 @@ EXIT_CODE=0
 
 # Check for run-clang-tidy, then run-clang-tidy-LLVM_VERSION, and
 # if we can't find those, then just loop over the files one at a time.
-if command -v run-clang-tidy > /dev/null 2>&1; then
-    run-clang-tidy -quiet ${THREADING_FLAG} -p ${BUILD_DIR} \
-                   ${ALL_SOURCE_FILES} || EXIT_CODE=1
-elif [ ! -z "${RUN_CLANG_TIDY_BIN}" ] && \
-         command -v ${RUN_CLANG_TIDY_BIN} > /dev/null 2>&1; then
-    ${RUN_CLANG_TIDY_BIN} -quiet ${THREADING_FLAG} -p ${BUILD_DIR} \
+RAN_CLANG_TIDY=1
```
Comment (wthrowe) [inline tools/ClangTidyAll.sh]: There's no reason for the values of this to be numbers, so how about 1 -> no and 0 -> yes to make the meaning clearer.  The test later will then have to change to `!=` (or `=`) to do string comparison.

File: tools/ClangTidyHash.sh [hunk 5]
```diff
@@ -57,14 +57,17 @@ EXIT_CODE=0
 
 # Check for run-clang-tidy, then run-clang-tidy-LLVM_VERSION, and
 # if we can't find those, then just loop over the files one at a time.
-if command -v run-clang-tidy > /dev/null 2>&1; then
-    run-clang-tidy -quiet ${THREADING_FLAG} -p ${BUILD_DIR} \
-                   ${MODIFIED_FILES[@]} || EXIT_CODE=1
-elif [ ! -z "${RUN_CLANG_TIDY_BIN}" ] && \
-         command -v ${RUN_CLANG_TIDY_BIN} > /dev/null 2>&1; then
-    ${RUN_CLANG_TIDY_BIN} -quiet ${THREADING_FLAG} -p ${BUILD_DIR} \
+RAN_CLANG_TIDY=1
```
Comment (wthrowe) [inline tools/ClangTidyHash.sh]: Same comment as in ClangTidyAll.sh

**Review by wthrowe — CHANGES_REQUESTED** [review]

File: tools/ClangTidyAll.sh (hunk 4 above)
Comment (nilsdeppe) [reply tools/ClangTidyAll.sh]: I have pushed a fixup that hopefully does what you imagined... Though not being well-versed in bash (comparisons especially still confuse me), please let me know if I got it wrong.

Comment (wthrowe) [reply tools/ClangTidyAll.sh]: You got it.  I'm fine with a squash.

(Comparisons are easy: you just learn the rules for perl and then do the opposite. And also remember that `=` is a synonym for `==` and both are in common use.  :P)

Comment (nilsdeppe) [reply tools/ClangTidyAll.sh]: Okay, thanks! I'll give @markscheel a chance to look at the fixups before I squash :+1: 

(haha, that's amazing :P )

**Review by markscheel — COMMENTED** [review]
Looks good, please squash!

**Comment by nilsdeppe** [comment]
Okay, rebased on develop and squashed. Thank you both for the reviews! :)

**Comment by kidder** [comment]
@markscheel @wthrowe please look at

**Review by wthrowe — DISMISSED** [review]

**Comment by nilsdeppe** [comment]
rebased on develop to resolve conflict and get renamed CI check

File: tools/ClangTidyHash.sh [hunk 6]
```diff
@@ -3,16 +3,22 @@
...
+    echo "Wrong number of arguments passed to ClangTidyHash.sh, $#"
     echo "Expecting BUILD_DIR SOURCE_DIR HASH"
+    echo "or BUILD_DIR SOURCE_DIR HASH NUM_THREADS_TO_RUN_ON"
     exit 1
 fi
 
-BUILD_DIR=$1
-SOURCE_DIR=$2
+BUILD_DIR=$(realpath $1)
+SOURCE_DIR=$(realpath $2)
```
Comment (wthrowe) [inline tools/ClangTidyHash.sh:14]: I missed this before, but I think realpath has portability issues.  Someone on OSX should verify that this works.  @kidder?

**Review by nilsvu — COMMENTED** [review]
I can confirm the file selection works on macOS. Haven't checked anything else.

**Review by wthrowe — APPROVED** [review]
Sounds good, then.

**Review by markscheel — APPROVED** [review]

**Pull request merged**